modules/core/shared/src/main/scala/io/circe/Error.scala

```diff
@@ -103,13 +103,7 @@ object DecodingFailure {
     pathToRoot: Option[PathToRoot],
     ops: Eval[List[CursorOp]]
   ) extends DecodingFailure(lazyReason) {
-    override final def pathToRootString: Option[String] =
-      if (pathToRoot == Some(PathToRoot.empty)) {
-        // For backwards compatibility. We'll make this more consistent in 0.15.x
-        Some("")
-      } else {
-        pathToRoot.orElse(PathToRoot.fromHistory(ops.value).toOption).map(_.asPathString)
-      }
+    override final def pathToRootString: Option[String] = Some(CursorOp.opsToPath(history))
```

Comment:
It's still work in progress.

I eventually understood the differences in performance with the tests you provided. I still don't know how to report errors clearly in some cases.

I would avoid trying to return meaningful errors for all cursor combinations, since some of them are not even used in the codebase.